### ISSUE TITLE
Using value pack in closure crashes compiler

### ISSUE THREAD

**Issue by jasonbobier** [apple/swift#67503]
This code crashes the compiler:

```swift
struct SyncedTypes<T1, T2> {
	let properties: [(T1, inout T2) -> Void]
	
	init<each V: Equatable>(_ pair: repeat (KeyPath<T1, each V>, WritableKeyPath<T2, each V>)) {
		var properties = [(T1, inout T2) -> Void]()

		repeat properties.append( { (source: T1, destination: inout T2) in
			destination[keyPath: each pair.1] = source[keyPath: each pair.0]
		})

		self.properties = properties
	}
}
```

**Environment**

Xcode: Version 15.0 beta 4 (15A5195m)

Deployment target: iOS 17.0

**Comment by AnthonyLatsis** [comment]
Reduced: 
```swift
func helper(_: () -> Void) {}

func test<each V>(args: repeat each V) {
  repeat helper({ _ = each args })
}
```

**Comment by xedin** [comment]
This is a know problem and there a dupe for it, I just need to find it…

**Comment by AnthonyLatsis** [comment]
@xedin #66917? The stack trace is different though.

**Comment by xedin** [comment]
There was an earlier issue as well.

**Comment by xedin** [comment]
@sophiapoirier Maybe you remember what the issue was?

**Comment by sophiapoirier** [comment]
> @sophiapoirier Maybe you remember what the issue was?

I am not aware of an existing issue tracking this, though I have hit this assertion recently while in progress working on another bug with you, but that was just faulty code in progress.

That said, the initial and reduced versions are both invalid, and with a recent build from main do produce an appropriate diagnostics for me. And a corrected version of the original code does successfully type-check:

```swift
struct SyncedTypes<T1, T2> {
  let properties: [(T1, inout T2) -> Void]
  
  init<each V: Equatable>(_ pair: repeat (KeyPath<T1, each V>, WritableKeyPath<T2, each V>)) {
    var properties = [(T1, inout T2) -> Void]()

    properties.append( { (source: T1, destination: inout T2) in
      repeat (destination[keyPath: (each pair).1] = source[keyPath: (each pair).0])
    })

    self.properties = properties
  }
}
```

**Comment by AnthonyLatsis** [comment]
@sophiapoirier Your corrected version is semantically different. It appends a single closure that does a pack assignment, whereas the original does a pack `append`. The diagnostic is still confusing though, it should probably say that applying `repeat` to a closure is not supported yet.

**Comment by jasonbobier** [comment]
Did this get fixed? I'm seeing an error generated now from both the code that I have above and the code that @AnthonyLatsis wrote.

**Comment by AnthonyLatsis** [comment]
The crash must have been fixed by #67164 (it does not revert cleanly but I got my example to crash again with the same stack dump after some shenanigans). Opened a new issue for the error: #69282.
